File: app/macos/hyperwhisper/Managers/AudioRecording/Streaming/Providers/XAIStreamingStrategy.swift
//
//  XAIStreamingStrategy.swift
//  hyperwhisper
//
//  Direct WebSocket streaming to xAI's speech-to-text API using the user's
//  own Grok/xAI API key.
//

import Foundation
import OSLog

final class XAIStreamingStrategy: StreamingProviderStrategy {
    private let logger = Logger(subsystem: "com.hyperwhisper.app", category: "XAIStreaming")
    private var committedTranscript = ""

    func buildWebSocketURL(config: StreamingSessionConfig) -> URL? {
        guard let apiKey = config.apiKey, !apiKey.isEmpty else {
            logger.error("Cannot build xAI URL: API key is missing")
            return nil
        }

        var components = URLComponents()
        components.scheme = "wss"
        components.host = "api.x.ai"
        components.path = "/v1/stt"

        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "sample_rate", value: "16000"),
            URLQueryItem(name: "encoding", value: "pcm"),
            URLQueryItem(name: "interim_results", value: "true"),
            URLQueryItem(name: "endpointing", value: "300")
        ]

        if let language = Self.supportedFormattingLanguage(for: config.language) {
            queryItems.append(URLQueryItem(name: "language", value: language))
            logger.info("xAI formatting language set: \(language, privacy: .public)")
        } else if let language = config.language, !language.isEmpty {
            logger.info("xAI formatting disabled: \(language, privacy: .public) is not in supported formatting set")
        }

        components.queryItems = queryItems
        return components.url
    }

    func buildWebSocketRequest(url: URL, config: StreamingSessionConfig) -> URLRequest? {
        guard let apiKey = config.apiKey, !apiKey.isEmpty else {
            logger.error("Cannot build xAI request: API key is missing")
            return nil
        }

        var request = URLRequest(url: url)
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        return request
    }

    func startMessages(config: StreamingSessionConfig) -> [URLSessionWebSocketTask.Message] {
        // Reset the committed-transcript baseline so a reconnect doesn't
        // prefix-strip xAI's fresh server-side transcript against stale
        // pre-disconnect state. xAI sends no startup config message, so the
        // session begins on socket open and we return no messages here.
        committedTranscript = ""
        return []
    }

    func encodeAudioChunk(_ pcmData: Data) -> URLSessionWebSocketTask.Message {
        .data(pcmData)
    }

    func parseMessage(_ text: String) -> StreamingProviderEvent? {
        guard let data = text.data(using: .utf8) else {
            logger.error("xAI parseMessage: failed to convert text to UTF-8 data")
            return nil
        }

        let message: XAIMessage
        do {
            message = try JSONDecoder().decode(XAIMessage.self, from: data)
        } catch {
            logger.warning("xAI parseMessage: failed to decode JSON: \(error.localizedDescription, privacy: .public)")
            return nil
        }

        switch message.type {
        case "transcript.created":
            logger.info("xAI transcript session created")
            return .sessionStarted(sessionId: nil)

        case "transcript.partial":
            guard let transcript = message.text, !transcript.isEmpty else { return nil }
            if message.is_final == true {
                guard let delta = committedDelta(from: transcript), !delta.isEmpty else {
                    return nil
                }
                return .finalTranscript(text: delta)
            }
            return .partialTranscript(text: transcript)

        case "transcript.done":
            if let transcript = message.text,
               let delta = committedDelta(from: transcript),
               !delta.isEmpty {
                return .finalTranscriptAndSessionComplete(
                    text: delta,
                    durationSeconds: message.duration ?? 0,
                    creditsUsed: 0
                )
            }
            return .sessionComplete(durationSeconds: message.duration ?? 0, creditsUsed: 0)

        case "error":
            return .error(message: message.message ?? "xAI streaming transcription failed")

        default:
            logger.debug("xAI unrecognized message type: \(message.type, privacy: .public)")
            return nil
        }
    }

    func stopSequence() -> [StreamingStopStep] {
        [
            .sendText(#"{"type":"audio.done"}"#),
            .waitForSessionComplete(timeout: 10.0),
            .closeWebSocket
        ]
    }

    var transcriptionProviderLabel: String { "xAI (Streaming)" }

    var supportsVocabulary: Bool { false }
}

private extension XAIStreamingStrategy {
    static let supportedFormattingLanguages: Set<String> = [
        "ar", "cs", "da", "de", "en", "es", "fa", "fil", "fr", "hi",
        "id", "it", "ja", "ko", "mk", "ms", "nl", "pl", "pt", "ro",
        "ru", "sv", "th", "tr", "vi"
    ]

    static let languageAliases: [String: String] = [
        "tl": "fil"
    ]

    static func supportedFormattingLanguage(for code: String?) -> String? {
        guard let raw = code?.trimmingCharacters(in: .whitespacesAndNewlines),
              !raw.isEmpty else { return nil }
        let lower = raw.lowercased()
        if lower == "auto" { return nil }
        let primary = lower.split(separator: "-").first.map(String.init) ?? lower
        let normalized = languageAliases[primary] ?? primary
        return supportedFormattingLanguages.contains(normalized) ? normalized : nil
    }

    func committedDelta(from transcript: String) -> String? {
        let normalized = transcript.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty else { return nil }

        if committedTranscript.isEmpty {
            committedTranscript = normalized
            return normalized
        }

        if normalized.hasPrefix(committedTranscript) {
            let suffix = normalized.dropFirst(committedTranscript.count)
                .trimmingCharacters(in: .whitespacesAndNewlines)
            committedTranscript = normalized
            return suffix.isEmpty ? nil : String(suffix)
        }

        if committedTranscript.hasPrefix(normalized) {
            return nil
        }

        committedTranscript += " " + normalized
        return normalized
    }
}

private struct XAIMessage: Decodable {
    let type: String
    let text: String?
    let is_final: Bool?
    let duration: Double?
    let message: String?
}

File: app/windows/HyperWhisper/Services/Streaming/XaiStreamingStrategy.cs
using System;
using System.Collections.Generic;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using HyperWhisper.Models;

namespace HyperWhisper.Services.Streaming;

public sealed class XaiStreamingStrategy : IStreamingProviderStrategy
{
    private static readonly HashSet<string> SupportedFormattingLanguages = new(StringComparer.OrdinalIgnoreCase)
    {
        "ar", "cs", "da", "de", "en", "es", "fa", "fil", "fr", "hi",
        "id", "it", "ja", "ko", "mk", "ms", "nl", "pl", "pt", "ro",
        "ru", "sv", "th", "tr", "vi"
    };

    private static readonly Dictionary<string, string> LanguageAliases = new(StringComparer.OrdinalIgnoreCase)
    {
        ["tl"] = "fil"
    };

    private string _committedTranscript = string.Empty;

    public string TranscriptionProviderLabel => "xAI (Streaming)";
    public bool SupportsVocabulary => false;
    public bool SessionStartsOnWebSocketOpen => false;
    public int AudioSampleRate => 16000;
    public IReadOnlyList<(byte[] Data, WebSocketMessageType Type)> GetStartMessages(StreamingSessionConfig config) => [];

    public Uri? BuildWebSocketUri(StreamingSessionConfig config)
    {
        if (string.IsNullOrWhiteSpace(config.ApiKey))
        {
            LoggingService.Warn("XaiStreamingStrategy: missing API key");
            return null;
        }

        var query = new List<string>
        {
            "sample_rate=16000",
            "encoding=pcm",
            "interim_results=true",
            "endpointing=300"
        };

        var language = SupportedFormattingLanguage(config.Language);
        if (!string.IsNullOrWhiteSpace(language))
        {
            query.Add($"language={Uri.EscapeDataString(language)}");
        }

        return new Uri($"wss://api.x.ai/v1/stt?{string.Join("&", query)}");
    }

    public void ConfigureWebSocket(ClientWebSocket webSocket, StreamingSessionConfig config)
    {
        if (!string.IsNullOrWhiteSpace(config.ApiKey))
        {
            webSocket.Options.SetRequestHeader("Authorization", $"Bearer {config.ApiKey}");
        }
    }

    public (byte[] Data, WebSocketMessageType Type) EncodeAudioChunk(byte[] pcmData) =>
        (pcmData, WebSocketMessageType.Binary);

    public StreamingProviderEvent? ParseMessage(string text)
    {
        try
        {
            var message = JsonSerializer.Deserialize<XaiMessage>(text);
            if (message == null) return null;

            return message.Type switch
            {
                "transcript.created" => new StreamingProviderEvent.SessionStarted(null),
                "transcript.partial" => ParseTranscriptPartial(message),
                "transcript.done" => ParseTranscriptDone(message),
                "error" => new StreamingProviderEvent.Error(message.Message ?? "xAI streaming transcription failed"),
                _ => null
            };
        }
        catch (Exception ex)
        {
            LoggingService.Warn($"XaiStreamingStrategy: failed to parse message: {ex.Message}");
            return null;
        }
    }

    public IReadOnlyList<StreamingStopStep> GetStopSequence() =>
    [
        TextStep("{\"type\":\"audio.done\"}"),
        new StreamingStopStep(StreamingStopAction.WaitForSessionComplete, WaitAfter: TimeSpan.FromSeconds(10)),
        new StreamingStopStep(StreamingStopAction.Close)
    ];

    public Task OnAudioSendOpportunityAsync(
        Func<byte[], WebSocketMessageType, CancellationToken, Task> webSocketSendAsync,
        CancellationToken cancellationToken
    )
    {
        return Task.CompletedTask;
    }

    private StreamingProviderEvent? ParseTranscriptPartial(XaiMessage message)
    {
        if (string.IsNullOrWhiteSpace(message.Text))
            return null;

        if (message.IsFinal == true)
        {
            var delta = CommittedDelta(message.Text);
            return string.IsNullOrWhiteSpace(delta)
                ? null
                : new StreamingProviderEvent.FinalTranscript(delta);
        }

        return new StreamingProviderEvent.PartialTranscript(message.Text);
    }

    private StreamingProviderEvent ParseTranscriptDone(XaiMessage message)
    {
        if (!string.IsNullOrWhiteSpace(message.Text))
        {
            var delta = CommittedDelta(message.Text);
            if (!string.IsNullOrWhiteSpace(delta))
            {
                return new StreamingProviderEvent.FinalTranscriptAndSessionComplete(
                    delta,
                    message.Duration ?? 0,
                    0
                );
            }
        }

        return new StreamingProviderEvent.SessionComplete(message.Duration ?? 0, 0);
    }

    private string? CommittedDelta(string transcript)
    {
        var normalized = transcript.Trim();
        if (normalized.Length == 0)
            return null;

        if (_committedTranscript.Length == 0)
        {
            _committedTranscript = normalized;
            return normalized;
        }

        if (normalized.StartsWith(_committedTranscript, StringComparison.Ordinal))
        {
            var suffix = normalized[_committedTranscript.Length..].Trim();
            _committedTranscript = normalized;
            return suffix.Length == 0 ? null : suffix;
        }

        if (_committedTranscript.StartsWith(normalized, StringComparison.Ordinal))
            return null;

        _committedTranscript += " " + normalized;
        return normalized;
    }

    private static string? SupportedFormattingLanguage(string? code)
    {
        if (string.IsNullOrWhiteSpace(code))
            return null;

        var normalized = code.Trim().ToLowerInvariant();
        if (normalized == "auto")
            return null;

        var separatorIndex = normalized.IndexOf('-');
        var primary = separatorIndex > 0 ? normalized[..separatorIndex] : normalized;
        var aliased = LanguageAliases.GetValueOrDefault(primary, primary);
        return SupportedFormattingLanguages.Contains(aliased) ? aliased : null;
    }

    private static StreamingStopStep TextStep(string json) =>
        new(StreamingStopAction.SendMessage, Encoding.UTF8.GetBytes(json), WebSocketMessageType.Text);

    private sealed class XaiMessage
    {
        [JsonPropertyName("type")]
        public string? Type { get; set; }

        [JsonPropertyName("text")]
        public string? Text { get; set; }

        [JsonPropertyName("is_final")]
        public bool? IsFinal { get; set; }

        [JsonPropertyName("duration")]
        public double? Duration { get; set; }

        [JsonPropertyName("message")]
        public string? Message { get; set; }
    }
}
